test/core/event_engine/fuzzing_event_engine/fuzzing_event_engine.h

```diff
@@ -48,62 +57,196 @@ class FuzzingEventEngine : public EventEngine {
-      override;
+      ABSL_LOCKS_EXCLUDED(mu_) override;
 
   ConnectionHandle Connect(OnConnectCallback on_connect,
                            const ResolvedAddress& addr,
                            const EndpointConfig& args,
-                           MemoryAllocator memory_allocator,
-                           Duration timeout) override;
+                           MemoryAllocator memory_allocator, Duration timeout)
+      ABSL_LOCKS_EXCLUDED(mu_) override;
 
-  bool CancelConnect(ConnectionHandle handle) override;
+  bool CancelConnect(ConnectionHandle handle) ABSL_LOCKS_EXCLUDED(mu_) override;
 
   bool IsWorkerThread() override;
 
   std::unique_ptr<DNSResolver> GetDNSResolver(
       const DNSResolver::ResolverOptions& options) override;
 
-  void Run(Closure* closure) override;
-  void Run(absl::AnyInvocable<void()> closure) override;
-  TaskHandle RunAfter(Duration when, Closure* closure) override;
-  TaskHandle RunAfter(Duration when,
-                      absl::AnyInvocable<void()> closure) override;
-  bool Cancel(TaskHandle handle) override;
+  void Run(Closure* closure) ABSL_LOCKS_EXCLUDED(mu_) override;
+  void Run(absl::AnyInvocable<void()> closure)
+      ABSL_LOCKS_EXCLUDED(mu_) override;
+  TaskHandle RunAfter(Duration when, Closure* closure)
+      ABSL_LOCKS_EXCLUDED(mu_) override;
+  TaskHandle RunAfter(Duration when, absl::AnyInvocable<void()> closure)
+      ABSL_LOCKS_EXCLUDED(mu_) override;
+  bool Cancel(TaskHandle handle) ABSL_LOCKS_EXCLUDED(mu_) override;
 
   using Time = std::chrono::time_point<FuzzingEventEngine, Duration>;
 
   Time Now() ABSL_LOCKS_EXCLUDED(mu_);
 
-  static void SetGlobalNowImplEngine(FuzzingEventEngine* engine)
-      ABSL_LOCKS_EXCLUDED(mu_);
-  static void UnsetGlobalNowImplEngine(FuzzingEventEngine* engine)
-      ABSL_LOCKS_EXCLUDED(mu_);
+  // Clear any global hooks installed by this event engine. Call prior to
+  // destruction to ensure no overlap between tests if constructing/destructing
+  // each test.
+  void UnsetGlobalHooks() ABSL_LOCKS_EXCLUDED(mu_);
 
  private:
+  // One pending task to be run.
   struct Task {
     Task(intptr_t id, absl::AnyInvocable<void()> closure)
         : id(id), closure(std::move(closure)) {}
     intptr_t id;
     absl::AnyInvocable<void()> closure;
   };
 
+  // Per listener information.
+  // We keep a shared_ptr to this, one reference held by the FuzzingListener
+  // Listener implementation, and one reference in the event engine state, so it
+  // may be iterated through and inspected - principally to discover the ports
+  // on which this listener is listening.
+  struct ListenerInfo {
+    ListenerInfo(
+        Listener::AcceptCallback on_accept,
+        absl::AnyInvocable<void(absl::Status)> on_shutdown,
+        std::unique_ptr<MemoryAllocatorFactory> memory_allocator_factory)
+        : on_accept(std::move(on_accept)),
+          on_shutdown(std::move(on_shutdown)),
```

Comment:
I don't see where on_shutdown is ever called. Am I missing that?

Reply:
in `~ListenerInfo`